### ISSUE TITLE
Async data transfer from CPU to GPU

### ISSUE THREAD

**Issue by Rhett-Ying** [dmlc/dgl#3294]
## 🚀 Feature

In some scenarios, data transfer from CPU to GPU could be the main part or even the bottleneck. we'd like to transfer required data in asynchronous way and beforehand. As a result, GPU/CPU could do other works during waiting for completion of data transfer. Namely, overlap and parallelism could be obtained then.

For now, we focus below 2 cases.
    Case 1: Graph is stored on GPU, but feature/label data leaves on CPU(probably data is too large to store on GPU). all operations including bathcing/sampling/model fwd/bwd runs on GPU. we call it as 'allG' case.
    Case 2: Graph and data lie on CPU, batching/sampling on CPU but model fwd/bwd runs on GPU. we call it as 'c2g'. In this case, we utilize both CPU and GPU computation resources to obtain better performance if possible.

In above cases, one major part is the data transfer from CPU to GPU which could be the bottleneck. This where async data transfer applies. below is the PoC results we've obtained on these cases.
<p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;" id="docs-internal-guid-5ca7ea4f-7fff-b934-5ef7-19d73d2ce485"><span style="font-size:12pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:700;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">perf results (//examples/pytorch/graphsage/train_sampling.py):</span></p><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:12pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:700;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">&nbsp;&nbsp;&nbsp; </span><span style="font-size:12pt;font-family:Arial;color:#000000;background-color:transparent;font-weight:700;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">allG: Case1 shown below.&nbsp; &nbsp; &nbsp; &nbsp; c2g: Case2 shown below.</span></p><div dir="ltr" style="margin-left:0pt;" align="left">

g4dn.4xlarge_T4/p3.2xlarge_V100 | allG/c2g | reddit_sync (s/epoch) | reddit_Async | ogbn-prod_sync | ogbn-pd_Async
-- | -- | -- | -- | -- | --
T4 | allG | 9.57 | 6.32 | 5.38 | 2.74
  | c2g | 12.06 | 6.4 | 10.17 | 6.59
V100 | allG | 9.08 | 5.46 | 9.21 | 5.04
  | c2g | 12.84 | 8.71 | 13.71 | 11.37

</div>

## Motivation
In some scenarios, GPU is idle and waiting for data from CPU. For better performance, we’d better prepare data beforehand and keep GPU busy.

## Pitch
We split one batch train into 2 separate stages. One(stage_1) is for batching/sampling/data_prepare/transfer, the other(stage_2) is for waiting_data and model forward/backward propagation. Each stage runs on separate python threads. As a result, each operation in different stages/threads has a chance to be overlapped and run in parallel

as for interfaces,
1. a new or improved NodeDataloader/EdgeDataloader wrapper to support run on sub-thread and transfer data asynchronously.
2. DGLGraph should be improved to be copied to pinned memory for realy async copy from host to device in cuda.

**Comment by Rhett-Ying** [comment]
Here's the target API I wanna achieve and I'd like to split it into smaller/independent sub-features. pls help review it. @jermainewang 

`class NodeDataLoader:
    def __init__(self, g, nids, block_sampler, device=None,
                  use_ddp=False, ddp_seed=0,
                  preload=0,
                  fetch_input={}, fetch_output={},
                  async_transfer=False,
                  **kwargs)
`
1. `preload`: enable preload feature for NodeDataLoader, this is mainly about booting a background thread for preloading.  Not depend on any other parts.
2. `g.to(non_blocking=True, streamID=0)`: this is mainly about add async transfer support for DGLGraph. API in python and C++ should be modified accordingly. Not depend on any other parts.
3. `fetch_input:{}, fetch_output:{}`: fetch required feat/label and attach them to generated blocks. Efficient index slicing should be applied here. But it does not depend on async transfer between cpu and gpu. Not depend on any other parts.
4. `async_transfer`: this will control whether to transfer feat/label/graph in synchronous way. As for the stream used for async transfer, we could utilize torch.cuda.Stream().cuda_stream in python which could be safely converted into cudaStream_t and used in C++. this part depends on Item2.
